### PR TITLE
Rpk considerations

### LINKED ISSUE #92: resolve duplicated descriptions in section Pinning of Raw Public Keys
To Be Completed further (TBD): Note, the above paragraphs are duplicated from the section {{survey}} so we may have to resolve this duplication.

### PR DESCRIPTION
**Pull request by mcr**
maybe close #92 
Is this enough detail on RPK pinning?

### CODE REVIEW

**Comment by mcr** [comment]
review from @elear too.

**Comment by mcr** [comment]
also #77

**Comment by petervanderstok** [comment]
Hi Michael,

I looked at th RPK text. For me it can be much shorter. (My opinion).
It is ony necessary to specify where the RPK is stored during transport; Arguments about not parsing the certificate seem obvious to me and certainly to someone who chooses to use RPK.

Security consideration belong in the security considerations section and not here.

Shorter names for the vocuvher attributes are indeed welcome. I was frightened away from using them.

cheerio and thanks,

Peter

File: constrained-voucher.mkd [hunk 1]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+The private key can be loaded in whatever format is appropriate and efficient for the hardware.
+(EDNOTE: are there lwig references that would make sense here?)
+
+## The Registrar Trust Anchor
+
+When the Pledge first connects to the Registrar, the connection to the Registrar is provisional, as explained in section 5.6.2 of {{I-D.ietf-anima-bootstrapping-keyinfra}}.
+The Registrar provides its public key in a TLSServerCertificate, and the Pledge uses that to validate that integrity of the (D)TLS connection, but it does not validate the identity of
+the provided certificate.
+
+If instead of using a (TLSServer)Certificate of type X509 (see section 4.4.2 of {{RFC8446}}),
```
Comment (EskoDijk) [inline constrained-voucher.mkd:536]: This I don't understand. If the Pledge connects with an IDevID certificate, why would the Registrar respond with an RPK as its server identity?   I thought the idea was that the Pledge can "scan" the Registrar server cert for the public-key part and use that for asserting proximity and, later on, validation against the pinned RPK in the Voucher.

I assume here we don't want to introduce another RPK method that impacts the Registrar in this way?

File: constrained-voucher.mkd [hunk 2]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+
+## The Voucher Request
+
+The Pledge can put the Registrar's public key into the proximity-registrar-subject-public-key-info
+(PRSPKI) field of the voucher-request.
+(The proximity-registrar-sha256-of-subject-public-key-info can also be used if the 32-bytes of hash turned out to be smaller than a typical ECDSA key.)
+
+As the format of the PRSPKI field is identical to the TLS Certificate RawPublicKey, no manipulation at all is needed to insert this into a voucher-request.
+
+## The Voucher Reply
```
Comment (EskoDijk) [inline constrained-voucher.mkd:547]: Reply -> Response

File: constrained-voucher.mkd [hunk 3]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+
+A returned voucher will have a pinned-domain-subject-public-key-info field with the identical key as was found in the proximity field above, as well as in the TLS connection.
+Validation of this is what takes the connection out of the provisional state (see {{I-D.ietf-anima-bootstrapping-keyinfra}} section 5.6.2).
+
+The voucher needs to be validated first.
+The Pledge needs to have a public key to validate the signature from the MASA on the voucher.
+
+Certificate chains may be included in the COSE-signed Constrained Vouchers described in this document, using the x5bag attribute from {{I-D.ietf-cose-x509}}.
```
Comment (EskoDijk) [inline constrained-voucher.mkd:555]: If possible we should leave this out! We only define, AFAIK, using the x5bag for the Registrar to send a bag of certs (including one chain) to MASA in a Voucher Request. So not for the BRSKI-EST protocol. And not in Vouchers.
To avoid confusion, let's remove - we are now in the RPK section so no need to say how the non-RPK protocol looks like.

File: constrained-voucher.mkd [hunk 4]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+However, this would require that the Pledge is able to form a certificate chain and validate it against a built-in Trust Anchor.
+
+A simpler method is that the MASA would sign the Voucher directly with a key pair for which the Pledge has the public key built-in to its firmware.
+That is, the MASA would use a degenerate PKI of level one.
+
+This has significant advantages on the Pledge, but presents security issues for the MASA, as it needs to keep a private online for signing purposes.
```
Comment (EskoDijk) [inline constrained-voucher.mkd:561]: From here on the content seems mostly security considerations; and these should be moved to that section. (Peter already commented something similar.)

File: constrained-voucher.mkd [hunk 5]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+
+The above mechanism would not help with loss or compromise if every key was loaded in the MASA.
+As the IDevID contains a pointer to the MASA, each Pledge can also refer the Registrar to a specific MASA (the one with the right key loaded), by changing the MASA URL encoded into the IDevID.
+This is a form of sharding: it may also have benefits when it comes to managing load when it is expected that many thousand of devices per second will attempt to onboard.
+
+The document {{I-D.richardson-anima-masa-considerations}} section 2.4 has more details.
+
+## Enrollment
+
+This document makes the use of the /crts call to retrieve an CSR Attributes object optional.
```
Comment (EskoDijk) [inline constrained-voucher.mkd:581]: /crts -> /att
also add "optional for the Pledge".

File: constrained-voucher.mkd [hunk 6]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+
+A Pledge that does not do this does not need to change the Certificate Signing Request (CSR) that it sends to the Registrar.
+The ASN.1 bits of the CSR can therefore be treated almost as a binary blob, signed with the IDevID.
```
Comment (EskoDijk) [inline constrained-voucher.mkd:584]: The CSR is signed with a private key from a newly generated keypair. The public key of that new keypair is the one to be included in the LDevID later on.
So the signed CSR can't be precreated because the Pledge would need a sizable supply of pre-created keypairs along with signed CSRs!  The Pledge doesn't know how often in its lifetime it will bootstrap or re-enroll, so in practice the Pledge needs to generate the new key on the fly I think.
Although: I can imagine a vendor that sells a Pledge with a *limit* on number of onboardings/enrolls.... not desirable from a customer point of view.

File: constrained-voucher.mkd [hunk 7]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+
+(EDNOTE: Can the signing even be done ahead of time?  Is there any live proof of possession?)
+
+The resulting LDevID that is received, if arranged to be paired with the existing private key used for the IDevID, can then be treated as a binary blob to be sent when authenticating other
```
Comment (EskoDijk) [inline constrained-voucher.mkd:588]: See my comment above; the "paired" part here is unclear to me since the keypair used for the LDevID differs from the IDevID keypair.

File: constrained-voucher.mkd [hunk 8]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+(D)TLS connections.
+
+XXX - how does the Pledge validate the Registrar in the future?
```
Comment (EskoDijk) [inline constrained-voucher.mkd:591]: In a previous discussion I thought we concluded that "BRSKI+EST", even in the constrained case with RPKs, is impossible to do if the Pledge cannot do cert chaining/validation.
Reason: for EST it needs to validate a Registrar (cmcRA) of the Domain by chaining a newly received Registrar cert (which is received when doing a DTLS connection to a new Registrar, or an updated cert of a Registrar living on the same Registrar IP address i.e. renewed identity )  to a known Trust Anchor Domain CA. 
So I would not go in the current draft into cert-less operation for the EST part.

File: constrained-voucher.mkd [hunk 9]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+
+XXX - if we the next step is to do EAP-TLS for WPA-Enteprise use, can we get a Raw Public Key for the server somewhere?
```
Comment (EskoDijk) [inline constrained-voucher.mkd:593]: Preferably remove this and make a Github issue out of it.

**Review by EskoDijk — CHANGES_REQUESTED** [review]
See my comments; if possible adapt text before merge but we can also merge in and then adapt.
Maybe first the scope discussion is needed: is our scope
1. Only define how RPK can be included (format) in Voucher/VoucherReq without details on how a protocol would use it?  (Peter voted for this)
2. Define the format and also how BRSKI would use this , e.g. for shorter Vouchers/VoucherReqs? (Esko's vote)   This leaves open future uses of the same RPK fields.
3. As 2. but also define how a Pledge that can't parse X.509v3 in detail, i.e. cannot do chaining, can do the BRSKI + EST protocols? (current PR text tries to go here, but I don't see yet how that would be possible.)

File: constrained-voucher.mkd (hunk 8 above)
Comment (mcr) [reply constrained-voucher.mkd:591]: When you say, "BRSKI+EST", do you mean to say that when using RPK, that the pledge can't enroll with EST?
There are a number of BRSKI use cases where the goal is to get a credential that can be used with an application server (but, not with other peer devices).  The application server can be thing-in-cloud, or can be some kind of ACE AS.
I have, however, also opened #99

Comment (EskoDijk) [reply constrained-voucher.mkd:591]: What I meant is that a Pledge that cannot do cert chaining/validation operations, also cannot do the EST part of the protocol we describe in this draft. (The protocol being the "CoAP/CBOR version of BRSKI followed by the CoAP version of EST".)  This is independent of whether an RPK of the Registrar is pinned or the full cert of the Registrar is pinned.
(A Domain CA cannot be pinned in case a Pledge cannot do cert chaining/validation operations, because the Pledge cannot verify the Registar then.)  See further comments in #99 . 

If the Pledge can do chaining/validation operations, there seems to be no problem - see also #99.  

I think BRSKI can carry any additional credentials as needed in the Voucher.

File: constrained-voucher.mkd (hunk 1 above)
Comment (mcr) [reply constrained-voucher.mkd:536]: Ack: how would the registrar know what it should respond with, this early in the protocol?

Comment (EskoDijk) [reply constrained-voucher.mkd:536]: Update: if I read RFC 7250, it seems clear that a client (PLedge) already signals in its first DTLS Client Hello handshake message that it supports RPK for the handshake and the server signals its support back in Server Hello. That seems to solve the issue.

So it would be good to keep the different cases in mind when writing the text:
1. Regular DTLS handshake based on IDevID/Registrar certs
2. Regular DTLS handshake based on IDevID/Registrar certs - but the Pledge cannot do PKIX validation and can only "extract" the RPK (SubjectPublicKeyInfo) structure from the Registrar's cert chain, discarding the rest.
3. RPK DTLS handshake per RFC 7250

For case 1/2 the benefit is that Registrar can get the MASA URI from the IDevID. For case 3 this isn't possible, so the Registrar must have a-priori a list of the Pledges to enroll and their public key identities along with info on which vendor/MASA each Pledge uses.

File: constrained-voucher.mkd [hunk 10]
```diff
@@ -520,6 +506,92 @@ One key difference is that the names of the leaves in the YANG does not have a m
...
+the provided certificate.
+
+If instead of using a (TLSServer)Certificate of type X509 (see section 4.4.2 of {{RFC8446}}),
+a RawPublicKey object is used, then the Pledge does not need to any decoding of the RF5280 format certifiate to get the public key.
+
+## The Voucher Request
+
+The Pledge can put the Registrar's public key into the proximity-registrar-subject-public-key-info
+(PRSPKI) field of the voucher-request.
+(The proximity-registrar-sha256-of-subject-public-key-info can also be used if the 32-bytes of hash turned out to be smaller than a typical ECDSA key.)
```
Comment (EskoDijk) [inline constrained-voucher.mkd:543]: Note: to apply the recent renaming operation to use "pubk" / "pubk-sha256" also on all the new text in this PR.

File: constrained-voucher.mkd (hunk 3 above)
Comment (EskoDijk) [reply constrained-voucher.mkd:555]: Update: I checked RFC 8366 and indeed the CMS structure there may include multiple certificates as needed for the Pledge, to build a chain of trust between the MASA's signing identity used and the pre-installed TA in the Pledge. Additionally the TA as used by the Pledge should be included in the CMS structure. 

In the constrained case, using COSE signing, typically no cert is included in the COSE structure but just the signature and (if really needed) a key identifier.  So to me it is still not clear why we would want to put any certificates in a Voucher in an x5bag at all ; given that we are now in the RPK section and thus one of the below conditions hold:

1) Pledge requires RPK-pinning and is unable to perform X.509 chaining
2) RPK-pinning is used to reduce size of Voucher  (putting certs in x5bag isn't going to reduce ... )
3) Pledge authenticates using only its RPK, Registrar does this too per RFC 7250, Registrar's RPK is pinned.

In conclusion I still don't see how this text on x5bag and cert chains would fit here.

File: constrained-voucher.mkd (hunk 6 above)
Comment (EskoDijk) [reply constrained-voucher.mkd:584]: Update: 802.1AR combined with BRSKI with RFC 7030, allows any of these Pledge behaviors:
1) All LDevIDs will use same keypair as IDevID  
2) Pledge stores some spare keypairs in secure memory, to use for LDevIDs  (limit on # of onboards/enrolls!). It may renew or rekey per the options 3/4 below.
3) Pledge generates new keypair for new LDevID some of the times (not new for renew, new for rekey) - Pledge decides itself or by some operator trigger whether to renew or rekey
4) Pledge generates new keypair for new LDevID every time - always rekey.

( (1) has the problem that cert revoking due to private key leakage becomes useless. The new cert will be insecure as well. )

So best to clarify that when a Pledge treats CSR as a "binary blob" this only works in a case 1) or 2) situation.

File: constrained-voucher.mkd (hunk 7 above)
Comment (EskoDijk) [reply constrained-voucher.mkd:588]: See above comments on "binary blob", can work but only in those specific cases.

File: constrained-voucher.mkd (hunk 8 above)
Comment (EskoDijk) [reply constrained-voucher.mkd:591]: Update: after our discussion in the call, I think the Pledge can do the EST part of the protocol (BRSKI-EST) in principle. It obtains an operational cert LDevID that it may use as a "binary blob" when doing a DTLS handshake later on with some peer device in the domain.  It can authenticate a peer device during a (new) DTLS handshake post-bootstrap, if the device/Pledge has been configured before that with identities of trusted peers. These identities can be validated by their RPK. The Pledge may either use 1) RFC 7250 RPK handshake; or 2) use normal cert handshake and it can extract the public key from the peer's (server) cert and compare that to its configured-peers TA list.

That said, #99 is still an issue so doing a minimal-Pledge approach puts limitations on how BRSKI can be used.
